platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/ServiceScopeValidator.java

```diff
@@ -52,24 +55,53 @@ public void whenRegistered(Class<? extends Annotation> annotation, Registration
     }
 
     private void validateScope(Class<?> serviceType) {
-        Class<? extends Scope> serviceScope = scopeOf(serviceType);
+        Class<? extends Scope>[] serviceScopes = scopeOf(serviceType);
 
-        if (serviceScope == null || ServiceScopeValidatorWorkarounds.shouldSuppressValidation(serviceType)) {
+        if (serviceScopes == null) {
             return;
         }
 
-        if (!scope.equals(serviceScope)) {
-            throw new IllegalArgumentException(invalidScopeMessage(serviceType, serviceScope));
+        if (serviceScopes.length == 0) {
+            throw new IllegalArgumentException(String.format("Service '%s' is declared with empty scope list", serviceType.getName()));
         }
+
+        if (!containsScope(scope, serviceScopes)) {
+            throw new IllegalArgumentException(invalidScopeMessage(serviceType, serviceScopes));
+        }
+    }
+
+    private static boolean containsScope(Class<? extends Scope> expectedScope, Class<? extends Scope>[] actualScopes) {
+        return Arrays.asList(actualScopes).contains(expectedScope);
```

Comment:
I would have gone with the refactoring I suggested before. It would have made the code more straightforward (a single `contains` call) and avoided this allocation (`asList`), although inlining plus escape analysis should be enough in this context 🤞

Reply:
Here is the follow-up:
- https://github.com/gradle/gradle/pull/27830